R/mle-finder.R

```diff
@@ -11,13 +11,82 @@ linear_gradient <- function(coef, design, outcome, noise_var = 1) {
   return(gradient)
 }
 
-linear_BFGS <- function(design, outcome, noise_var = 1) {
+MLE_BFGS_function <- function(loglik_function, gradient_function, design, outcome, ...) {
   initial <- rep(0, ncol(design))
-  MLE <- stats::optim(initial, linear_loglik, linear_gradient,
-    design = design, outcome = outcome, noise_var = 1,
-    control = list(fnscale = -1),
-    method = "BFGS"
+  BFGS_MLE <- stats::optim(initial, loglik_function, gradient_function,
+    design = design, outcome = outcome, control = list(fnscale = -1),
+    method = "BFGS",...
   )$par
-  return(MLE)
+  return(BFGS_MLE)
 }
 
+logit_loglik <- function(coef, design, outcome) {
+
+  n <- length(outcome)
+  prob <- expit_function(design %*% coef)
+  log_lik <- sum(outcome * log(prob) + (1 - outcome) * log(1 - prob))
+  return(log_lik)
+
+}
+
+logit_gradient <- function(coef, design, outcome) {
+
+  prob <- expit_function(design %*% coef)
+  gradient <- as.vector(t(outcome - prob) %*% design)
+
+  return(gradient)
+
+}
+
+
+logit_hessian <- function(coef, design, outcome){
+
+  prob <-  expit_function(design %*% coef)
+  W <- diag(c(prob * (1 - prob)))
+  hessian <- t(design) %*% W %*% design
+  return(hessian)
+}
+
+logit_newton <- function(design, outcome){
+
+  max_iteration <- 1000
+  iter <- 0
+
+  stop_indicator <- FALSE
+
+  coef_old <- rep(0, ncol(design))
+
+  loglik_old <- logit_loglik(coef_old, design, outcome)
+
+  while(!stop_indicator & iter <= max_iteration ){
+
+  prob_old <-  expit_function(design %*% coef_old)
+  hessian_old <- logit_hessian(coef_old, design, outcome)
+
+  coef_update <-  coef_old + solve(hessian_old, t(design) %*% (outcome-prob_old))
+  loglik_update <- logit_loglik(coef_update, design, outcome)
+  stop_indicator <- are_all_close(loglik_update, loglik_old)
+
+  loglik_old <- loglik_update
+  coef_old <- coef_update
+  iter <- iter+1
+
+  }
+  if(iter == max_iteration){
+    warning("The max iterations are reached!")
+  }
+  return(as.vector(coef_update))
+
+}
+
+
+expit_function <- function(x){
+  return(exp(x)/(1+exp(x)))
+}
+
+are_all_close <- function(v, w, abs_tol = 1e-6, rel_tol = 1e-6) {
```

Comment:
you can remove the version in helper.R since now that it's in the main package, you don't need it in the helper